euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Sort.java

```diff
@@ -0,0 +1,305 @@
+import cz.seznam.euphoria.core.client.flow.Flow;
+import cz.seznam.euphoria.core.client.functional.UnaryFunction;
+import cz.seznam.euphoria.core.client.graph.DAG;
+import cz.seznam.euphoria.core.client.io.Context;
+import cz.seznam.euphoria.core.client.operator.state.ListStorage;
+import cz.seznam.euphoria.core.client.operator.state.ListStorageDescriptor;
+import cz.seznam.euphoria.core.client.operator.state.State;
+import cz.seznam.euphoria.core.client.operator.state.StateFactory;
+import cz.seznam.euphoria.core.client.operator.state.StorageProvider;
+import cz.seznam.euphoria.core.client.util.Pair;
+import cz.seznam.euphoria.shaded.guava.com.google.common.base.Preconditions;
+import cz.seznam.euphoria.shaded.guava.com.google.common.collect.Lists;
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+@Derived(
+    state = StateComplexity.CONSTANT,
+    repartitions = 1
+)
+public class Sort<
+        IN, S extends Comparable<? super S>, W extends Window>
+    extends StateAwareWindowWiseSingleInputOperator<
+        IN, IN, IN, Integer, IN, W,
+    Sort<IN, S, W>> {
+  
+  private static final class Sorted<V>
+      extends State<V, V>
+      implements StateSupport.MergeFrom<Sorted<V>> {
+
+    @SuppressWarnings("unchecked")
+    static final ListStorageDescriptor SORT_STATE_DESCR =
+        ListStorageDescriptor.of("sort", (Class) Object.class);
+
+    final ListStorage<V> curr;
+    final Comparator<V> cmp;
+    
+    @SuppressWarnings("unchecked")
+    Sorted(Context<V> context, StorageProvider storageProvider, Comparator<V> cmp) {
+      super(context);
+      this.curr = (ListStorage<V>) storageProvider.getListStorage(SORT_STATE_DESCR);
+      this.cmp = cmp;
+    }
+    
+    @Override
+    public void add(V element) {
+      curr.add(element);
+    }
+    
+    @Override
+    public void flush() {
+      List<V> toSort = Lists.newArrayList(curr.get());
+      Collections.sort(toSort, cmp);
+      toSort.forEach(getContext()::collect);
+    }
+    
+    @Override
+    public void close() {
+      curr.clear();
+    }
+    
+    @Override
+    public void mergeFrom(Sorted<V> other) {
+      for (V v : other.curr.get()) {
+        add(v);
+      }
+    }
+  }
+
+  public static class OfBuilder {
+    private final String name;
+
+    OfBuilder(String name) {
+      this.name = name;
+    }
+
+    public <IN> ByBuilder<IN> of(Dataset<IN> input) {
+      return new ByBuilder<>(name, input);
+    }
+  }
+
+  public static class ByBuilder<IN> {
+    private final String name;
+    private final Dataset<IN> input;
+
+    ByBuilder(String name, Dataset<IN> input) {
+      this.name = requireNonNull(name);
+      this.input = requireNonNull(input);
+    }
+
+    public <S extends Comparable<? super S>> WindowByBuilder<IN, S> by(UnaryFunction<IN, S> sortByFn) {
```

Comment:
regarding myself: i'm fine with both. so `.by` is fully fine with me. just wanted to through the question onto the table.